render-ws-spark-client/src/main/java/org/janelia/render/client/spark/zspacing/ZPositionCorrectionClient.java

```diff
@@ -0,0 +1,206 @@
+package org.janelia.render.client.spark.zspacing;
+
+import com.beust.jcommander.ParametersDelegate;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.List;
+
+import org.apache.spark.SparkConf;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.janelia.alignment.spec.stack.StackId;
+import org.janelia.alignment.spec.stack.StackWithZValues;
+import org.janelia.render.client.ClientRunner;
+import org.janelia.render.client.parameter.CommandLineParameters;
+import org.janelia.render.client.parameter.MultiProjectParameters;
+import org.janelia.render.client.parameter.ZSpacingParameters;
+import org.janelia.render.client.spark.LogUtilities;
+import org.janelia.render.client.spark.pipeline.AlignmentPipelineParameters;
+import org.janelia.render.client.spark.pipeline.AlignmentPipelineStep;
+import org.janelia.render.client.zspacing.CrossCorrelationData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Spark client for generating z-layer cross correlation data and then using that data
+ * to estimate thickness for a range of z-layers in an aligned render stack.
+ *
+ * @author Eric Trautman
+ */
+public class ZPositionCorrectionClient
+        implements Serializable, AlignmentPipelineStep {
+
+    public static class Parameters extends CommandLineParameters {
+        @ParametersDelegate
+        public MultiProjectParameters multiProject = new MultiProjectParameters();
+
+        @ParametersDelegate
+        public ZSpacingParameters zSpacing = new ZSpacingParameters();
+
+        public org.janelia.render.client.zspacing.ZPositionCorrectionClient buildJavaClient(final StackWithZValues stackIdWithZValues,
+                                                                                            final int comparisonRange,
+                                                                                            final String runName,
+                                                                                            final boolean solveExisting)
+                throws IOException {
+
+            final org.janelia.render.client.zspacing.ZPositionCorrectionClient.Parameters jClientParameters =
+                    new org.janelia.render.client.zspacing.ZPositionCorrectionClient.Parameters();
+
+            jClientParameters.renderWeb.baseDataUrl = multiProject.baseDataUrl;
+
+            final StackId stackId = stackIdWithZValues.getStackId();
+            jClientParameters.renderWeb.owner = stackId.getOwner();
+            jClientParameters.renderWeb.project = stackId.getProject();
+            jClientParameters.stack = stackId.getStack();
+
+            jClientParameters.layerRange.minZ = stackIdWithZValues.getFirstZ();
+            jClientParameters.layerRange.maxZ = stackIdWithZValues.getLastZ();
```

Comment:
Something about this "set public fields two layers in" doesn't seem right to me. Is there a way of supplying those parameters during construction of the parameter object?

Reply:
Yeah - I've always been a little conflicted about my choice to publicly expose the command line parameters.  Although I think this usage is consistent with the overall public setup, I went ahead and added a constructor as you requested.